### PR TITLE
8321718: ProcessTools.executeProcess calls waitFor before logging

### PR DESCRIPTION
**Pull request by stefank**
There is some logging printed when tests spawns processes. This logging is triggered from calls to `OutputAnalyzer`, when it delegates calls to `LazyOutputBuffer`.

If we write code like this:
```
ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(...);
OutputAnalyzer output = new OutputAnalyzer(pb.start());
int exitValue = output.getExitValue();
```

We get the following logging:
```
[timestamp0] "Gathering output for process <pid>
[timestamp1] Waiting for completion for process <pid>
[timestamp2] Waiting for completion finished for process <pid>
```

The first line comes from the `OutputAnalyzer` constructor and the two other lines comes from the `getExitValue` call, which calls logs the above messages around the call to `waitFor`.

If instead write the code above as:
```
ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(...);
OutputAnalyzer output = ProcessTools.executeProcess(pb);
int exitValue = output.getExitValue();
```

We get the same logging, but timestamp1 and timestamp2 are almost the same. This happens because there's an extra call to `waitFor` inside `executeProcess`, but that `waitFor` does not have the "wait for" logging. So, instead we get the logging for the no-op `waitFor` inside `getExitValue`.

I would like to propose a small workaround to solve this. The workaround is that `executeProcess` delegates the `waitFor` call to the `LazyOutputBuffer` via `OutputAnalyzer`.  This is a small, limited workaround for this issue. Ideally I would like to extract the entire Process handling code out of LazyOutputBuffer and OutputAnalyzer, but the prototype for that rewrites all usages of `new OutputAnalyzer(pb.start())` and stretches far and wide in the test directories, so I'm starting out with this suggestion.

We can see of this patch by looking at the timestamps generated from the included test. Without the proposed workaround:

Without
```
testExecuteProcessExit
[2023-12-11T11:05:41.854579260Z] Gathering output for process 2547719
[2023-12-11T11:05:44.018335073Z] Waiting for completion for process 2547719
[2023-12-11T11:05:44.018851972Z] Waiting for completion finished for process 2547719

testExecuteProcessStdout
[2023-12-11T11:05:44.049509860Z] Gathering output for process 2547741
[2023-12-11T11:05:46.227768897Z] Waiting for completion for process 2547741
[2023-12-11T11:05:46.228021173Z] Waiting for completion finished for process 2547741


testNewOutputAnalyzerExit
[2023-12-11T11:05:46.231475003Z] Gathering output for process 2547782
[2023-12-11T11:05:46.231972095Z] Waiting for completion for process 2547782
[2023-12-11T11:05:48.421324123Z] Waiting for completion finished for process 2547782

testNewOutputAnalyzerStdout
[2023-12-11T11:05:48.424801269Z] Gathering output for process 2547929
```

With the proposed workaround:
```
testExecuteProcessExit
[2023-12-11T11:06:12.203633363Z] Gathering output for process 2550961
[2023-12-11T11:06:12.285826294Z] Waiting for completion for process 2550961
[2023-12-11T11:06:14.361912745Z] Waiting for completion finished for process 2550961

testExecuteProcessStdout
[2023-12-11T11:06:14.398685794Z] Gathering output for process 2550982
[2023-12-11T11:06:14.399518617Z] Waiting for completion for process 2550982
[2023-12-11T11:06:16.595986889Z] Waiting for completion finished for process 2550982

testNewOutputAnalyzerExit
[2023-12-11T11:06:16.602218373Z] Gathering output for process 2551067
[2023-12-11T11:06:16.603176801Z] Waiting for completion for process 2551067
[2023-12-11T11:06:18.793886942Z] Waiting for completion finished for process 2551067

testNewOutputAnalyzerStdout
[2023-12-11T11:06:18.796332176Z] Gathering output for process 2551172
```

See how the timestamp for "Waiting for completion for process" becomes "earlier" in the "executeProcess" cases.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321718](https://bugs.openjdk.org/browse/JDK-8321718): ProcessTools.executeProcess calls waitFor before logging (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [145e1eaa](https://git.openjdk.org/jdk/pull/17052/files/145e1eaaa0697f0d4fcbd34bfede2bc2ee60a708)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [3275136c](https://git.openjdk.org/jdk/pull/17052/files/3275136c02cb3f1dced28d01ab57fff2e5cc36be)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17052/head:pull/17052` \
`$ git checkout pull/17052`

Update a local copy of the PR: \
`$ git checkout pull/17052` \
`$ git pull https://git.openjdk.org/jdk.git pull/17052/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17052`

View PR using the GUI difftool: \
`$ git pr show -t 17052`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17052.diff">https://git.openjdk.org/jdk/pull/17052.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17052#issuecomment-1849983223)

### CODE REVIEW

File: test/lib/jdk/test/lib/process/OutputAnalyzer.java [hunk 1]
```diff
@@ -106,6 +106,14 @@ public OutputAnalyzer(String stdout, String stderr, int exitValue)
         buffer = OutputBuffer.of(stdout, stderr, exitValue);
     }
 
+    /**
+     * Delegate waitFor to the OutputBuffer. This ensures that
+     * the progress and timestmaps are logged correctly.
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/process/OutputAnalyzer.java:111]: Typo: timestmaps

**Review by dholmes-ora — APPROVED** [review]
Looks good.

Thanks

File: test/lib/jdk/test/lib/process/OutputAnalyzer.java (hunk 1 above)
Comment (stefank) [reply test/lib/jdk/test/lib/process/OutputAnalyzer.java]: Thanks. Updated

**Comment by stefank** [comment]
Thanks for reviewing! I'm considering removing the test before integrating, as it currently takes >8s to run and it was mainly used to show the difference between the implementations. Do you agree that it would be OK to remove the test?

**Comment by dholmes-ora** [comment]
> Do you agree that it would be OK to remove the test?

Sure that's fine.

**Review by jaikiran — APPROVED** [review]
Hello Stefan, these changes look good to me.

Like you note, the new test case in this PR, isn't needed, so can be removed.

**Comment by stefank** [comment]
Thanks for reviewing! I'll remove the test, merge, and will do some sanity checks before integrating.

**Pull request closed without merging**